syntax = "proto3";

package quintet.raft.rpc;

message PbReply {
  uint64 term = 1;
  bool ans = 2;
}

message PbLogEntry {
  uint64 term = 1;
  string opName = 2;
  string args = 3;
  uint64 prmIdx = 4;
}

message PbAppendEntriesMessage {
  uint64 term = 1;
  string leaderId = 2;
  uint64 prevLogIdx = 3;
  uint64 prevLogTerm = 4;
  repeated PbLogEntry logEntries = 5;
  uint64 commitIdx = 6;
}

message PbRequestVoteMessage {
  uint64 term = 1;
  string candidateId = 2;
  uint64 lastLogIdx = 3;
  uint64 lastLogTerm = 4;
}

service RaftRpc {
  rpc AppendEntries (PbAppendEntriesMessage) returns (PbReply) {}

  rpc RequestVote (PbRequestVoteMessage) returns (PbReply) {}
}

